libs/commons/Common.ml

```diff
@@ -99,6 +99,29 @@ let before_return f v =
   f v;
   v
 
+(*****************************************************************************)
+(* Composition/Control *)
+(*****************************************************************************)
+
+let protect ~finally work =
+  let prev_blocked = ref None in
+  let finally () =
+    (try
+       prev_blocked := Some (Unix.sigprocmask Unix.SIG_BLOCK [ Sys.sigalrm ])
+     with
+    | Invalid_argument _ -> (* TODO: on Windows ? *) ());
+    finally ()
+    (* If we unblock here, we could get a Timeout exception inside 'finally' ... *)
+  in
+  (* nosemgrep: no-fun-protect *)
+  let x = Fun.protect ~finally work in
+  (* If 'prev_blocked' is 'None' then nothing was blocked... if it's 'Some _' then
+   * we are not on Windows so future calls should not raise any exception. *)
+  !prev_blocked
+  |> Option.iter (fun prev_blocked ->
+         Unix.sigprocmask Unix.SIG_SETMASK prev_blocked |> ignore);
```

Comment:
What happens if a sigalarm is raised while it was masked? When you unmask it, it will be raised again? or the timeout is lost?